File: app/src/main/java/com/owncloud/android/datamodel/e2e/v1/decrypted/DecryptedMetadata.java
/*
 * Nextcloud - Android Client
 *
 * SPDX-FileCopyrightText: 2023 Tobias Kaminsky <tobias@kaminsky.me>
 * SPDX-FileCopyrightText: 2023 Nextcloud GmbH
 * SPDX-License-Identifier: AGPL-3.0-or-later OR GPL-2.0-only
 */
package com.owncloud.android.datamodel.e2e.v1.decrypted;

import java.util.Map;

public class DecryptedMetadata {
    transient
    private Map<Integer, String> metadataKeys; // outdated with v1.1
    private String metadataKey;
    private String checksum;
    private double version;

    @Override
    public String toString() {
        return String.valueOf(version);
    }

    public Map<Integer, String> getMetadataKeys() {
        return this.metadataKeys;
    }

    public String getMetadataKey() {
        if (metadataKey == null) {
            // fallback to old keys array
            return metadataKeys.get(0);
        }
        return metadataKey;
    }

    public double getVersion() {
        return this.version;
    }

    public void setMetadataKeys(Map<Integer, String> metadataKeys) {
        this.metadataKeys = metadataKeys;
    }

    public void setMetadataKey(String metadataKey) {
        this.metadataKey = metadataKey;
    }

    public void setVersion(double version) {
        this.version = version;
    }

    public String getChecksum() {
        return checksum;
    }

    public void setChecksum(String checksum) {
        this.checksum = checksum;
    }
}

File: app/src/main/java/com/owncloud/android/datamodel/e2e/v2/decrypted/DecryptedMetadata.kt
/*
 * Nextcloud - Android Client
 *
 * SPDX-FileCopyrightText: 2023 Tobias Kaminsky <tobias@kaminsky.me>
 * SPDX-FileCopyrightText: 2023 Nextcloud GmbH
 * SPDX-License-Identifier: AGPL-3.0-or-later OR GPL-2.0-only
 */
package com.owncloud.android.datamodel.e2e.v2.decrypted

import com.owncloud.android.utils.EncryptionUtils

data class DecryptedMetadata(
    val keyChecksums: MutableList<String> = mutableListOf(),
    val deleted: Boolean = false,
    var counter: Long = 0,
    val folders: MutableMap<String, String> = mutableMapOf(),
    val files: MutableMap<String, DecryptedFile> = mutableMapOf(),
    @Transient
    var metadataKey: ByteArray = EncryptionUtils.generateKey()
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as DecryptedMetadata

        if (keyChecksums != other.keyChecksums) return false
        if (deleted != other.deleted) return false
        if (counter != other.counter) return false
        if (folders != other.folders) return false
        if (files != other.files) return false

        return true
    }

    override fun hashCode(): Int {
        var result = keyChecksums.hashCode()
        result = 31 * result + deleted.hashCode()
        result = 31 * result + counter.hashCode()
        result = 31 * result + folders.hashCode()
        result = 31 * result + files.hashCode()
        return result
    }
}
